ext/openssl/ossl_pkey_rsa.c

```diff
@@ -628,6 +628,173 @@ ossl_rsa_to_public_key(VALUE self)
+
+    md_ctx = EVP_MD_CTX_new();
+    if (!md_ctx)
+        ossl_raise(ePKeyError, "EVP_MD_CTX_new");
+
+    pkey_ctx = EVP_PKEY_CTX_new(pkey, NULL);
+    if (!pkey_ctx) {
+        error = 1;
+        goto err;
+    }
+
+    if (EVP_DigestSignInit(md_ctx, &pkey_ctx, md, NULL, pkey) != 1) {
+        error = 2;
+        goto err;
+    }
+
+    EVP_PKEY_CTX_set_rsa_padding(pkey_ctx, RSA_PKCS1_PSS_PADDING);
+
+    if (EVP_DigestSignUpdate(md_ctx, RSTRING_PTR(data), RSTRING_LEN(data)) != 1) {
+        error = 3;
+        goto err;
+    }
+
+    if (EVP_DigestSignFinal(md_ctx, (unsigned char *)RSTRING_PTR(signature), &buf_len) != 1) {
+        error = 4;
+	    goto err;
+    }
+
+    rb_str_set_len(signature, buf_len);
+
+ err:
+    EVP_MD_CTX_destroy(md_ctx);
+    if (error) {
+        ossl_raise(ePKeyError, NULL);
+    }
+
+    return signature;
+}
+
+/*
+ *  call-seq:
+ *      pkey.verify_pss(digest, signature, data) -> String
+ *
+ * Probabilistic Signature Scheme for RSA verify().
+ *
+ * To verify the +String+ +signature+, +digest+ and +data+ must be provided
+ * to re-compute the message digest. See +sign_pss+ for details about
+ * +digest+ and +data+.
+ * The return value is +true+ if the
+ * signature is valid, +false+ otherwise. A PKeyError is raised should errors
+ * occur.
+ * Any previous state of the +Digest+ instance is irrelevant to the validation
+ * outcome, the digest instance is reset to its initial state during the
+ * operation.
+ *
+ * == Example
+ *   data = 'Sign me!'
+ *   digest = OpenSSL::Digest::SHA256.new
+ *   pkey = OpenSSL::PKey::RSA.new(2048)
+ *   signature = pkey.sign(digest, data)
+ *   pub_key = pkey.public_key
+ *   puts pub_key.verify_pss(digest, signature, data) # => true
+ */
+static VALUE
+ossl_rsa_verify_pss(VALUE self, VALUE digest, VALUE signature, VALUE data)
+{
+    EVP_PKEY *pkey;
+    EVP_PKEY_CTX *pkey_ctx;
+    const EVP_MD *md;
+    EVP_MD_CTX *md_ctx;
+    int result, error;
+    void *pkey_ptr;
+    const BIGNUM *n, *e;
+
+    error = 0;
+
+    GetPKeyRSA(self, pkey);
+    md = GetDigestPtr(digest);
+    StringValue(signature);
+    StringValue(data);
+
+    pkey_ptr = EVP_PKEY_get0(pkey);
+    RSA_get0_key(pkey_ptr, &n, &e, NULL);
+    if (!(n && e))
+        ossl_raise(ePKeyError, "missing public RSA key");
+
+    md_ctx = EVP_MD_CTX_new();
+    if (!md_ctx) {
+        ossl_raise(ePKeyError, "EVP_MD_CTX_new");
+    }
+
+    pkey_ctx = EVP_PKEY_CTX_new(pkey, NULL);
+    if (!pkey_ctx) {
+        error = 1;
+        goto err;
+    }
+
+    if (EVP_DigestVerifyInit(md_ctx, &pkey_ctx, md, NULL, pkey) != 1) {
+        error = 2;
+        goto err;
+    }
+
+    EVP_PKEY_CTX_set_rsa_padding(pkey_ctx, RSA_PKCS1_PSS_PADDING);
+
+    if (EVP_DigestVerifyUpdate(md_ctx, RSTRING_PTR(data), RSTRING_LEN(data)) != 1) {
+        error = 3;
+        goto err;
+    }
+
+    result = EVP_DigestVerifyFinal(md_ctx, (unsigned char *)RSTRING_PTR(signature), RSTRING_LENINT(signature));
```

Comment:
RSTRING_LENINT() raises a RangeError if a too long String is passed as the signature, and `md_ctx` will leak. It is a fault in the original code... I'll fix soon.